File: src/main/java/com/example/unitconverterapp/MainActivity.java
package com.example.unitconverterapp;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private Spinner spinnerFrom, spinnerTo;
    private EditText editTextValue;
    private Button buttonConvert;
    private TextView textViewResult;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        spinnerFrom = findViewById(R.id.spinnerFrom);
        spinnerTo = findViewById(R.id.spinnerTo);
        editTextValue = findViewById(R.id.editTextValue);
        buttonConvert = findViewById(R.id.buttonConvert);
        textViewResult = findViewById(R.id.textViewResult);

        //
        String[] units = {
                "Inch", "Foot", "Yard", "Mile",
                "Centimeter", "Meter", "Kilometer",
                "Pound", "Ounce", "Ton",
                "Gram", "Kilogram", "Celsius", "Fahrenheit", "Kelvin"
        };

        ArrayAdapter<String> adapter = new ArrayAdapter<>(
                this,
                R.layout.spinner_main_item,
                units
        );
        adapter.setDropDownViewResource(R.layout.spinner_item);
        spinnerFrom.setAdapter(adapter);
        spinnerTo.setAdapter(adapter);

        buttonConvert.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String fromUnit = spinnerFrom.getSelectedItem().toString();
                String toUnit = spinnerTo.getSelectedItem().toString();


                if (editTextValue.getText().toString().trim().isEmpty()) {
                    Toast.makeText(MainActivity.this, "Please enter the value you want to convert!", Toast.LENGTH_SHORT).show();
                    return;
                }

                double inputValue;
                try {
                    inputValue = Double.parseDouble(editTextValue.getText().toString().trim());
                } catch (NumberFormatException e) {
                    Toast.makeText(MainActivity.this, "The input is invalid, please enter the number!", Toast.LENGTH_SHORT).show();
                    return;
                }

                if (fromUnit.equals(toUnit)) {
                    textViewResult.setText("Same unit, no conversion required:" + inputValue);
                    return;
                }

                double result = convertUnits(fromUnit, toUnit, inputValue);

                textViewResult.setText(String.valueOf(result));
            }
        });
    }

    private double convertUnits(String fromUnit, String toUnit, double value) {

        double baseValue;

        if (isLengthUnit(fromUnit)) {
            baseValue = toMeter(fromUnit, value);
            return fromMeter(toUnit, baseValue);
        }

        if (isWeightUnit(fromUnit)) {
            baseValue = toKilogram(fromUnit, value);
            return fromKilogram(toUnit, baseValue);
        }

        if (isTemperatureUnit(fromUnit) && isTemperatureUnit(toUnit)) {
            return convertTemperature(fromUnit, toUnit, value);
        }

        Toast.makeText(this, "Conversion between types is not supported at this time!", Toast.LENGTH_SHORT).show();
        return 0;
    }

    private boolean isLengthUnit(String unit) {
        return unit.equals("Inch") || unit.equals("Foot") || unit.equals("Yard") || unit.equals("Mile")
                || unit.equals("Centimeter") || unit.equals("Meter") || unit.equals("Kilometer");
    }

    private boolean isWeightUnit(String unit) {
        return unit.equals("Pound") || unit.equals("Ounce") || unit.equals("Ton")
                || unit.equals("Gram") || unit.equals("Kilogram");
    }

    private boolean isTemperatureUnit(String unit) {
        return unit.equals("Celsius") || unit.equals("Fahrenheit") || unit.equals("Kelvin");
    }

    private double toMeter(String fromUnit, double value) {
        switch (fromUnit) {
            case "Inch":
                // 1 inch = 2.54 cm,0.0254 m
                return value * 0.0254;
            case "Foot":
                // 1 foot = 30.48 cm,0.3048 m
                return value * 0.3048;
            case "Yard":
                // 1 yard = 0.9144 m
                return value * 0.9144;
            case "Mile":
                // 1 mile = 1609.34 m
                return value * 1609.34;
            case "Centimeter":
                // 1 cm = 0.01 m
                return value * 0.01;
            case "Kilometer":
                // 1 km = 1000 m
                return value * 1000.0;
            case "Meter":
            default:
                return value;
        }
    }

    private double fromMeter(String toUnit, double meterValue) {
        switch (toUnit) {
            case "Inch":
                return meterValue / 0.0254;
            case "Foot":
                return meterValue / 0.3048;
            case "Yard":
                return meterValue / 0.9144;
            case "Mile":
                return meterValue / 1609.34;
            case "Centimeter":
                return meterValue * 100;
            case "Kilometer":
                return meterValue / 1000.0;
            case "Meter":
            default:
                return meterValue;
        }
    }

    private double toKilogram(String fromUnit, double value) {
        switch (fromUnit) {
            case "Pound":
                // 1 pound = 0.453592 kg
                return value * 0.453592;
            case "Ounce":
                // 1 ounce = 28.3495 g = 0.0283495 kg
                return value * 0.0283495;
            case "Ton":
                // 1 ton = 907.185 kg
                return value * 907.185;
            case "Gram":
                // 1 g = 0.001 kg
                return value * 0.001;
            case "Kilogram":
            default:
                return value;
        }
    }

    private double fromKilogram(String toUnit, double kgValue) {
        switch (toUnit) {
            case "Pound":
                return kgValue / 0.453592;
            case "Ounce":
                return kgValue / 0.0283495;
            case "Ton":
                return kgValue / 907.185;
            case "Gram":
                return kgValue * 1000.0;
            case "Kilogram":
            default:
                return kgValue;
        }
    }

    private double convertTemperature(String fromUnit, String toUnit, double value) {
        double celsiusValue;

        // Step1: from -> Celsius
        switch (fromUnit) {
            case "Celsius":
                celsiusValue = value;
                break;
            case "Fahrenheit":
                // C = (F - 32)/1.8
                celsiusValue = (value - 32) / 1.8;
                break;
            case "Kelvin":
                // C = K - 273.15
                celsiusValue = value - 273.15;
                break;
            default:
                celsiusValue = value;
        }

        // Step2: Celsius -> toUnit
        switch (toUnit) {
            case "Celsius":
                return celsiusValue;
            case "Fahrenheit":
                // F = C * 1.8 + 32
                return celsiusValue * 1.8 + 32;
            case "Kelvin":
                // K = C + 273.15
                return celsiusValue + 273.15;
            default:
                return celsiusValue;
        }
    }
}

File: src/main/java/com/example/unitconverterapp/MainActivity.kt
package com.example.unitconverterapp

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.input.KeyboardType
import com.example.unitconverterapp.ui.theme.CustomTypography


class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            MaterialTheme(
                typography = CustomTypography
            ) {
                UnitConverterApp()
            }
        }
    }
}

@Composable
fun UnitConverterApp() {
    var inputValue by remember { mutableStateOf("") }
    var sourceUnit by remember { mutableStateOf("Inch") }
    var targetUnit by remember { mutableStateOf("Centimeter") }
    var result by remember { mutableStateOf("") }

    val units = listOf("Inch", "Foot", "Yard", "Mile", "Centimeter", "Meter", "Kilometer",
        "Pound", "Ounce", "Ton", "Kilogram", "Gram",
        "Celsius", "Fahrenheit", "Kelvin")

    Column(
        modifier = Modifier.padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp)
    ) {
        Text(text = "Unit-Converter", style = MaterialTheme.typography.headlineLarge)

        OutlinedTextField(
            value = inputValue,
            onValueChange = { inputValue = it },
            label = { Text("Enter Value" , color = Color(0xFF1E90FF)) },
            keyboardOptions = KeyboardOptions.Default.copy(keyboardType = KeyboardType.Number),
            colors = TextFieldDefaults.colors(
                focusedIndicatorColor = Color(0xFF1E90FF),
                unfocusedIndicatorColor = Color.Gray,
                cursorColor = Color(0xFF1E90FF)
            )
        )

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(10.dp)
        ){
            UnitDropdownMenu(
                label = "From",
                selectedUnit = sourceUnit,
                units = units
            ) { sourceUnit = it }

            UnitDropdownMenu(
                label = "To",
                selectedUnit = targetUnit,
                units = units
            ) { targetUnit = it }
        }

        Button(onClick = {
            result = try {
                val input = inputValue.toDouble()
                convertUnits(input, sourceUnit, targetUnit).toString()
            } catch (e: Exception) {
                "Invalid Input"
            }
        },
            colors = ButtonDefaults.buttonColors(
                containerColor = Color(0xFF1E90FF)
            )
        ) {
            Text(text = "CONVERT" , color = Color.White)
        }

        Text(text = "Result: ", style = MaterialTheme.typography.headlineLarge)

        OutlinedTextField(
            value = result,
            onValueChange = {},
            label = { Text("Converted value " , color = Color(0xFF1E90FF))},
            readOnly = true,
            modifier = Modifier.fillMaxWidth(),
            colors = TextFieldDefaults.colors(
                    focusedIndicatorColor = Color(0xFF1E90FF), 
                    unfocusedIndicatorColor = Color.Gray,
                    cursorColor = Color(0xFF1E90FF)
            )
        )
    }
}

@Composable
fun UnitDropdownMenu(label: String, selectedUnit: String, units: List<String>, onUnitSelected: (String) -> Unit) {
    var expanded by remember { mutableStateOf(false) }
    Box {
        OutlinedButton(onClick = { expanded = true }) {
            Text("$label: $selectedUnit")
        }
        DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            units.forEach { unit ->
                DropdownMenuItem(
                    text = { Text(unit) },
                    onClick = {
                    onUnitSelected(unit)
                    expanded = false
                }
                )
            }
        }
    }
}

fun convertUnits(value: Double, from: String, to: String): Double {
    val lengthConversions = mapOf(
        "Inch" to 2.54, "Foot" to 30.48, "Yard" to 91.44, "Mile" to 160934.0, "Centimeter" to 1.0, "Meter" to 100.0, "Kilometer" to 100000.0
    )
    val weightConversions = mapOf(
        "Pound" to 0.453592, "Ounce" to 0.0283495, "Ton" to 907.185, "Kilogram" to 1.0, "Gram" to 0.001
    )

    return when {
        from == to -> value
        from in lengthConversions && to in lengthConversions ->
            value * (lengthConversions[from]!! / lengthConversions[to]!!)
        from in weightConversions && to in weightConversions ->
            value * (weightConversions[from]!! / weightConversions[to]!!)
        from == "Celsius" && to == "Fahrenheit" -> (value * 1.8) + 32
        from == "Fahrenheit" && to == "Celsius" -> (value - 32) / 1.8
        from == "Celsius" && to == "Kelvin" -> value + 273.15
        from == "Kelvin" && to == "Celsius" -> value - 273.15
        else -> throw IllegalArgumentException("Invalid Conversion")
    }
}

@Preview(showBackground = true)
@Composable
fun PreviewApp() {
    UnitConverterApp()
}

